File: src/conventions/pipeline/observation/observation-log-builder.ts
/**
 * Observation log builder — constructs a CommittedStep[] from per-step data.
 *
 * Single-pass O(n): iterates through steps once, threading kernel state.
 * The caller (strategy adapter) provides per-step registers and arbitration
 * results captured during protocol replay.
 *
 * **Temporal ordering:** The log contains observations for all PAST bids.
 * The current bid being evaluated is NOT in the log — its CommittedStep
 * is appended AFTER arbitration resolves it.
 */

import type { Seat, Call } from "../../../engine/types";
import type { CommittedStep, NegotiationState } from "../../core/committed-step";
import { INITIAL_NEGOTIATION } from "../../core/committed-step";
import type { PipelineResult } from "../pipeline-types";
import type { MachineRegisters } from "../../core/module-surface";
import { buildCommittedStep } from "./committed-step-builder";

/** Input for one auction step. Provided by the strategy adapter. */
export interface ObservationLogStep {
  readonly actor: Seat;
  readonly call: Call;
  readonly registers: MachineRegisters;
  readonly pipelineResult: PipelineResult | null;
}

/**
 * Build a CommittedStep[] from per-step auction data.
 *
 * Threads kernel state through the loop: step N's prevKernel is
 * step N-1's stateAfter (or INITIAL_NEGOTIATION for step 0).
 */
export function buildObservationLog(
  steps: readonly ObservationLogStep[],
): readonly CommittedStep[] {
  const log: CommittedStep[] = [];
  let prevKernel: NegotiationState = INITIAL_NEGOTIATION;

  for (const step of steps) {
    const committed = buildCommittedStep(
      step.actor,
      step.call,
      step.pipelineResult,
      prevKernel,
      step.registers,
    );
    log.push(committed);
    prevKernel = committed.stateAfter;
  }

  return log;
}

File: src-tauri/crates/bridge-conventions/src/pipeline/observation/observation_log_builder.rs
//! Observation log builder — constructs a CommittedStep[] from per-step data.
//!
//! Mirrors TS from `pipeline/observation/observation-log-builder.ts`.

use bridge_engine::types::{Call, Seat};

use crate::pipeline::observation::committed_step::{
    initial_negotiation, ClaimRef, CommittedStep, CommittedStepStatus,
};
use crate::pipeline::observation::negotiation_extractor::compute_kernel_delta;
use crate::pipeline::observation::normalize_intent::normalize_intent;
use crate::pipeline::types::PipelineResult;
use crate::types::negotiation::NegotiationState;

/// Input for one auction step.
pub struct ObservationLogStep {
    pub actor: Seat,
    pub call: Call,
    pub state_after: NegotiationState,
    pub pipeline_result: Option<PipelineResult>,
}

/// Build a CommittedStep vec from per-step auction data.
///
/// Threads kernel state through the loop: step N's prev_kernel is
/// step N-1's state_after (or INITIAL_NEGOTIATION for step 0).
pub fn build_observation_log(steps: &[ObservationLogStep]) -> Vec<CommittedStep> {
    let mut log = Vec::with_capacity(steps.len());
    let mut prev_kernel = initial_negotiation();

    for step in steps {
        let resolved_claim = extract_claim_ref(&step.pipeline_result);
        let public_actions = extract_public_obs(&step.pipeline_result);
        let negotiation_delta = compute_kernel_delta(&prev_kernel, &step.state_after);
        let status = derive_status(&step.pipeline_result);

        let committed = CommittedStep {
            actor: step.actor,
            call: step.call.clone(),
            resolved_claim,
            public_actions,
            negotiation_delta,
            state_after: step.state_after.clone(),
            status,
        };
        prev_kernel = committed.state_after.clone();
        log.push(committed);
    }

    log
}

fn extract_claim_ref(result: &Option<PipelineResult>) -> Option<ClaimRef> {
    let result = result.as_ref()?;
    let carrier = result.selected.as_ref()?;
    let proposal = &carrier.encoded.proposal;

    Some(ClaimRef {
        module_id: proposal.module_id.clone(),
        meaning_id: proposal.meaning_id.clone(),
        semantic_class_id: proposal.semantic_class_id.clone(),
        source_intent: proposal.source_intent.clone(),
    })
}

fn extract_public_obs(
    result: &Option<PipelineResult>,
) -> Vec<crate::types::bid_action::BidAction> {
    match result {
        Some(r) => match &r.selected {
            Some(carrier) => normalize_intent(&carrier.encoded.proposal.source_intent),
            None => Vec::new(),
        },
        None => Vec::new(),
    }
}

fn derive_status(result: &Option<PipelineResult>) -> CommittedStepStatus {
    match result {
        None => CommittedStepStatus::OffSystem,
        Some(r) => {
            if r.selected.is_some() {
                CommittedStepStatus::Resolved
            } else if !r.truth_set.is_empty() {
                CommittedStepStatus::Ambiguous
            } else {
                CommittedStepStatus::OffSystem
            }
        }
    }
}
